Struct fields behind multi-level pointers are half-broken: materialized structs panic in the write-back, nil chains are silently skipped, and recursive pointer types (`type P *P`) never terminate

## Summary

The per-field pointer handling in `processWith` is designed for exactly one
pointer level, and every path through it misbehaves once a field is `**T` or
deeper:

- the dereference loop materializes a nil pointer only when its immediate
  pointee is a struct, and otherwise skips the field:
  https://github.com/sethvargo/go-envconfig/blob/bdd371d05608e34392063f07659e2618f1e90dc6/envconfig.go#L466-L483
- the write-back closure assumes the origin field is a pointer directly to
  the materialized struct, and `reflect.Set`s the struct's address into it:
  https://github.com/sethvargo/go-envconfig/blob/bdd371d05608e34392063f07659e2618f1e90dc6/envconfig.go#L452-L464
- `processField` materializes pointer levels until it reaches a non-pointer,
  assuming that always happens:
  https://github.com/sethvargo/go-envconfig/blob/bdd371d05608e34392063f07659e2618f1e90dc6/envconfig.go#L862-L868

The write-back panic was noted in passing in #141 as out of scope of the
recursive-struct check; this issue files it properly, together with the rest
of the multi-level pointer surface it belongs to.

## What happens today, verified against v1.4.2

For a non-recursive struct type `Inner` with one `env` field:

- `P **Inner` holding a pointer to a nil `*Inner`: **panic** in the
  write-back. `setNilStruct` materializes the struct one level down but
  writes its `*Inner` address back into the `**Inner` field:
  `panic: reflect.Set: value of type *main.Inner is not assignable to type
  **main.Inner` (from `envconfig.go:461` via the call at `:529`).
- `P **Inner` nil, untagged: **silently skipped**, while a nil `*Inner`
  right next to it is materialized and traversed. One extra `*` turns a
  decoded field into a no-op with no error.
- `Sub **Inner` tagged `env:",prefix=SUB_"`: **ErrPrefixNotStruct** ("prefix
  is only valid on struct types"), while the same tag on `*Inner` works. The
  field falls off the struct path onto the leaf path, where a prefix is
  invalid.
- `P **Inner` tagged `env:",noinit"`: fails with `P: missing key`. The leaf
  path does not swallow `ErrMissingKey` for the empty key the way the
  struct branch does
  (https://github.com/sethvargo/go-envconfig/blob/bdd371d05608e34392063f07659e2618f1e90dc6/envconfig.go#L550-L553
  versus
  https://github.com/sethvargo/go-envconfig/blob/bdd371d05608e34392063f07659e2618f1e90dc6/envconfig.go#L495-L498).
- `P **Inner` fully populated: decodes correctly (the loop dereferences
  through non-nil levels and the struct is processed in place). Multi-level
  pointers to non-structs (`**string`) also work at any depth, nil or not,
  because `processField` materializes through the field itself; the test
  suite pins both (`pointer_pointer_string`, `no_overwrite/pointers_pointers`).

So the supported surface is data-dependent: populated chains decode, nil
chains silently do not, and the half-populated chain panics.

Separately, a named self-referential pointer type never reaches a value to
decode into, and nothing guards the two loops that walk pointer chains:

- `type P *P` with a value pointing at itself spins the dereference loop
  forever: a hard CPU hang, not a stack overflow, so `recover()` never
  fires. No struct is in the cycle, so the `ErrRecursiveStruct` check from
  #141 does not apply (verified: it hangs identically with that fix in
  place).
- A **nil** `P` behind an `env` tag reaches the pointer materialization in
  `processField`, which allocates a fresh level forever:
  `fatal error: out of memory` (from `envconfig.go:865`, reached via
  `:586`). The same happens for collection elements, e.g. `[]P` with a
  one-item value, via the per-element `processField` call at `:967`,
  even though the field itself is not a pointer.

## Affected versions

v1.4.2 and `main` (currently identical). None of this is new: the one-level
nil skip dates to the initial commit, and the write-back with its
single-level assumption dates to the nil-struct-pointer initialization work
(#43, first released in v0.5.0).

## Minimal reproductions

All outputs below are from v1.4.2.

**Write-back panic:**

```go
package main

import (
	"context"
	"fmt"

	"github.com/sethvargo/go-envconfig"
)

type Inner struct {
	Value string `env:"VALUE"`
}

type Outer struct {
	P **Inner
}

func main() {
	ctx := context.Background()

	var nilInner *Inner
	target := &Outer{P: &nilInner}
	err := envconfig.Process(ctx, target)
	fmt.Println("err:", err) // never reached
}
```

```
panic: reflect.Set: value of type *main.Inner is not assignable to type **main.Inner

goroutine 1 [running]:
reflect.Value.assignTo(...)
reflect.Value.Set(...)
github.com/sethvargo/go-envconfig.processWith.func1(...)
	.../envconfig.go:461
github.com/sethvargo/go-envconfig.processWith(...)
	.../envconfig.go:529
```

**Silent skip and the prefix flip:**

```go
package main

import (
	"context"
	"fmt"

	"github.com/sethvargo/go-envconfig"
)

type Inner struct {
	Value string `env:"VALUE"`
}

type Single struct {
	P *Inner
}

type Double struct {
	P **Inner
}

type PrefixSingle struct {
	Sub *Inner `env:", prefix=SUB_"`
}

type PrefixDouble struct {
	Sub **Inner `env:", prefix=SUB_"`
}

func main() {
	ctx := context.Background()
	lookuper := envconfig.MapLookuper(map[string]string{
		"VALUE":     "set",
		"SUB_VALUE": "set",
	})

	var s Single
	err := envconfig.ProcessWith(ctx, &envconfig.Config{Target: &s, Lookuper: lookuper})
	fmt.Printf("single-level:  err=%v P=%v", err, s.P)
	if s.P != nil {
		fmt.Printf(" P.Value=%q", s.P.Value)
	}
	fmt.Println()

	var d Double
	err = envconfig.ProcessWith(ctx, &envconfig.Config{Target: &d, Lookuper: lookuper})
	fmt.Printf("double-level:  err=%v P=%v\n", err, d.P)

	var ps PrefixSingle
	err = envconfig.ProcessWith(ctx, &envconfig.Config{Target: &ps, Lookuper: lookuper})
	fmt.Printf("prefix-single: err=%v Sub=%v", err, ps.Sub)
	if ps.Sub != nil {
		fmt.Printf(" Sub.Value=%q", ps.Sub.Value)
	}
	fmt.Println()

	var pd PrefixDouble
	err = envconfig.ProcessWith(ctx, &envconfig.Config{Target: &pd, Lookuper: lookuper})
	fmt.Printf("prefix-double: err=%v Sub=%v\n", err, pd.Sub)
}
```

```
single-level:  err=<nil> P=&{set} P.Value="set"
double-level:  err=<nil> P=<nil>
prefix-single: err=<nil> Sub=&{set} Sub.Value="set"
prefix-double: err=prefix is only valid on struct types Sub=<nil>
```

**Dereference-loop hang:**

```go
package main

import (
	"context"
	"fmt"

	"github.com/sethvargo/go-envconfig"
)

type P *P

type SelfPtr struct {
	Value string `env:"VALUE"`
	Link  P
}

func main() {
	ctx := context.Background()

	var p P
	p = &p
	err := envconfig.Process(ctx, &SelfPtr{Link: p})
	fmt.Println("err:", err) // never reached
}
```

Never returns; one core pinned until killed.

The unbounded-allocation variants are the same program with, respectively, a
nil `Link P` tagged `env:"LINK"` (and `LINK` set to anything), and a
`Links []P` field tagged `env:"LINKS"` with a one-item value. Each pointer
level the loop materializes stays reachable from the field, so the chain
grows until memory gives out and the process dies with an unrecoverable
`fatal error: out of memory` (output below captured under a 1 GiB
address-space cap to make the failure quick):

```
runtime: out of memory: cannot allocate 4194304-byte block (309886976 in use)
fatal error: out of memory

github.com/sethvargo/go-envconfig.processField(...)
	.../envconfig.go:865
github.com/sethvargo/go-envconfig.processWith(...)
	.../envconfig.go:586
```

## Proposed fix

Make pointer depth irrelevant to decoding semantics, and reject the shapes
that cannot terminate:

1. Materialize the struct at the end of a **nil pointer chain of any
   depth** (today: only depth one), and have the write-back target the
   pointer the traversal started from, rebuilding the levels in between,
   instead of assuming the origin is a single pointer-to-struct. This fixes
   the panic, the silent skip, the prefix flip, and the `noinit` "missing
   key" error in one stroke: `**Inner` then behaves exactly like `*Inner`
   for initialization, `noinit`, prefixes, and write-back, and populated
   chains keep decoding in place as they do today.
2. Reject pointer types whose dereference chain never reaches a non-pointer
   type (`type P *P`, mutually recursive pairs) with a new sentinel in the
   `ErrRecursiveStruct` family, checked both at the field level and in
   `processField` for collection elements. This is the same reasoning as
   #141 one layer down: no chain of such a type ever reaches a value to
   decode into, so the shape is an authoring error, and a deterministic
   type-level rejection beats a hang. The check is two `Kind` calls for
   ordinary pointer fields; the cycle walk only runs for
   pointer-to-pointer shapes.

Two behavior changes worth calling out honestly:

- A recursive **struct** type behind a nil `**T` currently decodes because
  the nil chain is skipped (pinned by `recursive/pointer_to_pointer_nil_ok`
  from #142, which documented the then-current boundary). With
  materialization reaching any depth, the descent re-enters the type and
  rejects it with `ErrRecursiveStruct`, exactly like the populated chain
  next to it in the table. That makes the #142 rejection fully type-level
  rather than data-dependent, which is the argument #142 itself made for
  rejecting the populated variant.
- Nil (or finite) values of recursive **pointer** types are currently
  skipped when untagged; they become an explicit error. Given any non-nil
  value hangs the process and any tagged nil value OOMs it, a fail-fast
  error seems strictly better than data-dependent silence.

The alternative I considered and dropped: rejecting multi-level pointers to
structs outright. That would break currently-working populated `**T`
decodes, and it would be inconsistent with multi-level pointers to
non-structs, which have always decoded through `processField` at any depth
(`pointer_pointer_string` pins this).

--- a/envconfig.go
+++ b/envconfig.go
@@ -97,6 +97,7 @@ const (
 	ErrNotStruct          = internalError("input must be a struct")
 	ErrPrefixNotStruct    = internalError("prefix is only valid on struct types")
 	ErrPrivateField       = internalError("cannot parse private fields")
+	ErrRecursivePointer   = internalError("pointer type is recursive")
 	ErrRecursiveStruct    = internalError("struct type is recursive")
 	ErrRequiredAndDefault = internalError("field cannot be required and have a default value")
 	ErrUnknownOption      = internalError("unknown option")
@@ -449,33 +450,55 @@ func processWith(ctx context.Context, c *Config, path map[reflect.Type]bool) err
 		required := structRequired || opts.Required
 
 		isNilStructPtr := false
+		var nilPtrOrigin reflect.Value
 		setNilStruct := func(v reflect.Value) {
-			origin := e.Field(i)
 			if isNilStructPtr {
-				empty := reflect.New(origin.Type().Elem()).Interface()
+				empty := reflect.New(v.Type().Elem()).Interface()
 
 				// If a struct (after traversal) equals to the empty value, it means
 				// nothing was changed in any sub-fields. With the noinit opt, we skip
 				// setting the empty value to the original struct pointer (keep it nil).
 				if !reflect.DeepEqual(v.Interface(), empty) || !noInit {
-					origin.Set(v)
+					// The origin is the nil pointer the traversal was started from.
+					// It can sit several pointer levels away from the struct, so
+					// rebuild the levels in between before writing back.
+					for v.Type() != nilPtrOrigin.Type() {
+						p := reflect.New(v.Type())
+						p.Elem().Set(v)
+						v = p
+					}
+					nilPtrOrigin.Set(v)
 				}
 			}
 		}
 
 		// Initialize pointer structs.
 		pointerWasSet := false
+		if ef.Kind() == reflect.Pointer && ef.Type().Elem().Kind() == reflect.Pointer {
+			// A named pointer type can dereference to itself (type P *P), directly
+			// or through other named pointer types. No chain of such a type ever
+			// reaches a value to decode into, so dereferencing one below (or
+			// materializing one in processField) would never terminate.
+			if _, ok := nonPointerType(ef.Type()); !ok {
+				return fmt.Errorf("%s: %s: %w", tf.Name, ef.Type(), ErrRecursivePointer)
+			}
+		}
 		for ef.Kind() == reflect.Pointer {
 			if ef.IsNil() {
-				if ef.Type().Elem().Kind() != reflect.Struct {
-					// This is a nil pointer to something that isn't a struct, like
-					// *string. Move along.
+				base, _ := nonPointerType(ef.Type()) // cannot cycle, checked above
+				if base.Kind() != reflect.Struct {
+					// This is a nil pointer chain to something that isn't a struct,
+					// like *string or **string. Move along; processField
+					// materializes those through the field itself.
 					break
 				}
 
 				isNilStructPtr = true
-				// Use an empty struct of the type so we can traverse.
-				ef = reflect.New(ef.Type().Elem()).Elem()
+				nilPtrOrigin = ef
+				// Use an empty struct of the base type so we can traverse; the
+				// write-back in setNilStruct rebuilds any pointer levels between
+				// the origin and the struct.
+				ef = reflect.New(base).Elem()
 			} else {
 				pointerWasSet = true
 				ef = ef.Elem()
@@ -853,6 +876,28 @@ func processAsDecoder(ctx context.Context, v string, ef reflect.Value) (bool, er
 	return imp, err
 }
 
+// nonPointerType returns the first non-pointer type reached by successively
+// dereferencing t, and reports whether one exists. Named pointer types can
+// dereference to themselves (type P *P), directly or through other named
+// pointer types; such a chain never reaches a non-pointer type. The two
+// cursors advance at different speeds, so on a cyclic chain they eventually
+// meet, without allocating a set of visited types.
+func nonPointerType(t reflect.Type) (reflect.Type, bool) {
+	slow, fast := t, t
+	for fast.Kind() == reflect.Pointer {
+		fast = fast.Elem()
+		if fast.Kind() != reflect.Pointer {
+			return fast, true
+		}
+		fast = fast.Elem()
+		slow = slow.Elem()
+		if slow == fast {
+			return nil, false
+		}
+	}
+	return fast, true
+}
+
 func processField(ctx context.Context, v string, ef reflect.Value, delimiter, separator string, noInit bool) error {
 	// If the input value is empty and initialization is skipped, do nothing.
 	if v == "" && noInit {
@@ -860,6 +905,14 @@ func processField(ctx context.Context, v string, ef reflect.Value, delimiter, se
 	}
 
 	// Handle pointers and uninitialized pointers.
+	if ef.Type().Kind() == reflect.Pointer && ef.Type().Elem().Kind() == reflect.Pointer {
+		// Collection elements do not pass through the field-level check in
+		// processWith, so reject recursive pointer types here too, before the
+		// materialization below chases the chain.
+		if _, ok := nonPointerType(ef.Type()); !ok {
+			return fmt.Errorf("%s: %w", ef.Type(), ErrRecursivePointer)
+		}
+	}
 	for ef.Type().Kind() == reflect.Pointer {
 		if ef.IsNil() {
 			ef.Set(reflect.New(ef.Type().Elem()))

--- a/envconfig.go
+++ b/envconfig.go
@@ -450,7 +450,7 @@ func processWith(ctx context.Context, c *Config, path map[reflect.Type]bool) err
 		required := structRequired || opts.Required
 
 		isNilStructPtr := false
-		var nilPtrOrigin reflect.Value
+		var nilPtrOrigin, nilPtrChain reflect.Value
 		setNilStruct := func(v reflect.Value) {
 			if isNilStructPtr {
 				empty := reflect.New(v.Type().Elem()).Interface()
@@ -459,15 +459,7 @@ func processWith(ctx context.Context, c *Config, path map[reflect.Type]bool) err
 				// nothing was changed in any sub-fields. With the noinit opt, we skip
 				// setting the empty value to the original struct pointer (keep it nil).
 				if !reflect.DeepEqual(v.Interface(), empty) || !noInit {
-					// The origin is the nil pointer the traversal was started from.
-					// It can sit several pointer levels away from the struct, so
-					// rebuild the levels in between before writing back.
-					for v.Type() != nilPtrOrigin.Type() {
-						p := reflect.New(v.Type())
-						p.Elem().Set(v)
-						v = p
-					}
-					nilPtrOrigin.Set(v)
+					nilPtrOrigin.Set(nilPtrChain)
 				}
 			}
 		}
@@ -495,10 +487,18 @@ func processWith(ctx context.Context, c *Config, path map[reflect.Type]bool) err
 
 				isNilStructPtr = true
 				nilPtrOrigin = ef
-				// Use an empty struct of the base type so we can traverse; the
-				// write-back in setNilStruct rebuilds any pointer levels between
-				// the origin and the struct.
-				ef = reflect.New(base).Elem()
+				// Materialize a detached pointer chain that keeps the field's
+				// declared type at every level, initializing each cell from its
+				// own element type the way processField's pointer loop does, and
+				// traverse the empty struct at its end. setNilStruct attaches
+				// the whole chain to the origin with a single Set.
+				nilPtrChain = reflect.New(ef.Type().Elem())
+				cell := nilPtrChain.Elem()
+				for cell.Kind() == reflect.Pointer {
+					cell.Set(reflect.New(cell.Type().Elem()))
+					cell = cell.Elem()
+				}
+				ef = cell
 			} else {
 				pointerWasSet = true
 				ef = ef.Elem()

--- a/envconfig_test.go
+++ b/envconfig_test.go
@@ -312,13 +312,29 @@ type reusedLeaf struct {
 	Value string `env:"VALUE"`
 }
 
-// doublePointerNode's self-reference sits behind a pointer to a pointer: the
-// descent materializes only pointers directly to structs, so a nil
-// **doublePointerNode is skipped rather than materialized.
+// doublePointerNode's self-reference sits behind a pointer to a pointer:
+// materialization reaches the struct at the end of a pointer chain of any
+// depth, so the type re-enters the descent path whether the chain is nil or
+// populated.
 type doublePointerNode struct {
 	Value string `env:"VALUE"`
 	Next  **doublePointerNode
 }
+
+// pointerChainLeaf sits at the end of the multi-level pointer chains in the
+// multi_pointer tests.
+type pointerChainLeaf struct {
+	Value string `env:"VALUE"`
+}
+
+// selfPointer dereferences to itself: no chain of it ever reaches a
+// non-pointer type to decode into.
+type selfPointer *selfPointer
+
+// mutualPointerA and mutualPointerB dereference to each other.
+type mutualPointerA *mutualPointerB
+
+type mutualPointerB *mutualPointerA
 
 func TestProcessWithConfigArgument(t *testing.T) {
 	t.Parallel()
@@ -3352,16 +3368,17 @@ func TestProcessWith(t *testing.T) {
 			}),
 		},
 		{
-			// A nil pointer-to-pointer is skipped, not materialized: the
-			// descent never re-enters the type, so the shape keeps decoding.
-			name:   "recursive/pointer_to_pointer_nil_ok",
+			// Materialization reaches the struct at the end of a pointer chain
+			// of any depth, so a recursive type behind a nil pointer-to-pointer
+			// re-enters the descent path exactly like the populated chain
+			// below: the rejection is type-level, not data-dependent.
+			name:   "recursive/pointer_to_pointer_nil",
 			target: &doublePointerNode{},
-			exp: &doublePointerNode{
-				Value: "v",
-			},
 			lookuper: MapLookuper(map[string]string{
 				"VALUE": "v",
 			}),
+			err:    ErrRecursiveStruct,
+			errMsg: "Next: envconfig.doublePointerNode: struct type is recursive",
 		},
 		{
 			// A populated pointer-to-pointer chain re-enters the type, so it
@@ -3399,6 +3416,203 @@ func TestProcessWith(t *testing.T) {
 				"A_VALUE": "a",
 				"B_VALUE": "b",
 			}),
+		},
+
+		// Multi-level pointers
+		{
+			// https://github.com/sethvargo/go-envconfig/issues/143
+			// A nil pointer chain of any depth ending at a struct is
+			// materialized and written back, wired through every level.
+			name: "multi_pointer/materializes_nil_chain",
+			target: &struct {
+				P **pointerChainLeaf
+			}{},
+			exp: &struct {
+				P **pointerChainLeaf
+			}{
+				P: ptrTo(ptrTo(pointerChainLeaf{Value: "v"})),
+			},
+			lookuper: MapLookuper(map[string]string{
+				"VALUE": "v",
+			}),
+		},
+		{
+			name: "multi_pointer/materializes_nil_chain_deep",
+			target: &struct {
+				P ***pointerChainLeaf
+			}{},
+			exp: &struct {
+				P ***pointerChainLeaf
+			}{
+				P: ptrTo(ptrTo(ptrTo(pointerChainLeaf{Value: "v"}))),
+			},
+			lookuper: MapLookuper(map[string]string{
+				"VALUE": "v",
+			}),
+		},
+		{
+			// Without noinit, a nil chain is initialized even when nothing was
+			// set, matching the single-level behavior.
+			name: "multi_pointer/initializes_empty_chain",
+			target: &struct {
+				P **pointerChainLeaf
+			}{},
+			exp: &struct {
+				P **pointerChainLeaf
+			}{
+				P: ptrTo(ptrTo(pointerChainLeaf{})),
+			},
+			lookuper: MapLookuper(nil),
+		},
+		{
+			// https://github.com/sethvargo/go-envconfig/issues/143
+			// A nil pointer partway down a populated chain: the write-back
+			// goes to that pointer, not to the field, which used to panic in
+			// reflect.Set.
+			name: "multi_pointer/inner_nil_write_back",
+			target: &struct {
+				P **pointerChainLeaf
+			}{
+				P: ptrTo((*pointerChainLeaf)(nil)),
+			},
+			exp: &struct {
+				P **pointerChainLeaf
+			}{
+				P: ptrTo(ptrTo(pointerChainLeaf{Value: "v"})),
+			},
+			lookuper: MapLookuper(map[string]string{
+				"VALUE": "v",
+			}),
+		},
+		{
+			// noinit holds across the whole chain: nothing was set, so the
+			// field stays nil at every level.
+			name: "multi_pointer/noinit_keeps_chain_nil",
+			target: &struct {
+				P **pointerChainLeaf `env:",noinit"`
+			}{},
+			exp: &struct {
+				P **pointerChainLeaf `env:",noinit"`
+			}{},
+			lookuper: MapLookuper(nil),
+		},
+		{
+			// noinit with a populated outer pointer and nothing set: the nil
+			// inner pointer is left alone.
+			name: "multi_pointer/noinit_keeps_inner_nil",
+			target: &struct {
+				P **pointerChainLeaf `env:",noinit"`
+			}{
+				P: ptrTo((*pointerChainLeaf)(nil)),
+			},
+			exp: &struct {
+				P **pointerChainLeaf `env:",noinit"`
+			}{
+				P: ptrTo((*pointerChainLeaf)(nil)),
+			},
+			lookuper: MapLookuper(nil),
+		},
+		{
+			// A fully populated chain decodes in place, no write-back needed.
+			name: "multi_pointer/populated_chain_in_place",
+			target: &struct {
+				P **pointerChainLeaf
+			}{
+				P: ptrTo(ptrTo(pointerChainLeaf{})),
+			},
+			exp: &struct {
+				P **pointerChainLeaf
+			}{
+				P: ptrTo(ptrTo(pointerChainLeaf{Value: "v"})),
+			},
+			lookuper: MapLookuper(map[string]string{
+				"VALUE": "v",
+			}),
+		},
+		{
+			// A prefix descends through any pointer depth; it used to be
+			// rejected with ErrPrefixNotStruct beyond one level.
+			name: "multi_pointer/prefix_descends_chain",
+			target: &struct {
+				Sub **pointerChainLeaf `env:", prefix=SUB_"`
+			}{},
+			exp: &struct {
+				Sub **pointerChainLeaf `env:", prefix=SUB_"`
+			}{
+				Sub: ptrTo(ptrTo(pointerChainLeaf{Value: "v"})),
+			},
+			lookuper: MapLookuper(map[string]string{
+				"SUB_VALUE": "v",
+			}),
+		},
+
+		// Recursive pointer types
+		{
+			// A self-referential pointer type never reaches a value to decode
+			// into, so it is rejected up front, nil or not. A non-nil chain
+			// used to spin the dereference loop forever.
+			name: "recursive_pointer/nil",
+			target: &struct {
+				Link selfPointer
+			}{},
+			lookuper: MapLookuper(nil),
+			err:      ErrRecursivePointer,
+			errMsg:   "Link: envconfig.selfPointer: pointer type is recursive",
+		},
+		{
+			// https://github.com/sethvargo/go-envconfig/issues/143
+			name: "recursive_pointer/self_pointing_value",
+			target: func() any {
+				var p selfPointer
+				p = &p
+				return &struct {
+					Link selfPointer
+				}{
+					Link: p,
+				}
+			}(),
+			lookuper: MapLookuper(nil),
+			err:      ErrRecursivePointer,
+			errMsg:   "Link: envconfig.selfPointer: pointer type is recursive",
+		},
+		{
+			// Mutually recursive pointer types are the same cycle one step
+			// apart.
+			name: "recursive_pointer/mutual",
+			target: &struct {
+				Link mutualPointerA
+			}{},
+			lookuper: MapLookuper(nil),
+			err:      ErrRecursivePointer,
+			errMsg:   "Link: envconfig.mutualPointerA: pointer type is recursive",
+		},
+		{
+			// A tagged nil field of a recursive pointer type used to reach the
+			// pointer materialization in processField and allocate without
+			// bound.
+			name: "recursive_pointer/tagged_nil",
+			target: &struct {
+				Link selfPointer `env:"LINK"`
+			}{},
+			lookuper: MapLookuper(map[string]string{
+				"LINK": "x",
+			}),
+			err:    ErrRecursivePointer,
+			errMsg: "Link: envconfig.selfPointer: pointer type is recursive",
+		},
+		{
+			// Collection elements decode through processField and its pointer
+			// materialization directly, without passing the field-level check,
+			// so the element type is rejected there.
+			name: "recursive_pointer/slice_element",
+			target: &struct {
+				Links []selfPointer `env:"LINKS"`
+			}{},
+			lookuper: MapLookuper(map[string]string{
+				"LINKS": "a",
+			}),
+			err:    ErrRecursivePointer,
+			errMsg: "envconfig.selfPointer: pointer type is recursive",
 		},
 	}
 

--- a/envconfig_test.go
+++ b/envconfig_test.go
@@ -327,6 +327,13 @@ type pointerChainLeaf struct {
 	Value string `env:"VALUE"`
 }
 
+// namedLeafPointer and namedDoubleLeafPointer are named pointer types in the
+// chain: the write-back must preserve the field's declared type at every
+// level rather than rebuilding the chain from unnamed pointer types.
+type namedLeafPointer *pointerChainLeaf
+
+type namedDoubleLeafPointer **pointerChainLeaf
+
 // selfPointer dereferences to itself: no chain of it ever reaches a
 // non-pointer type to decode into.
 type selfPointer *selfPointer
@@ -3543,6 +3550,53 @@ func TestProcessWith(t *testing.T) {
 			},
 			lookuper: MapLookuper(map[string]string{
 				"SUB_VALUE": "v",
+			}),
+		},
+		{
+			// A named pointer type as the field type: the attached value must
+			// carry the declared type, which only assignability permits (an
+			// identity-based rebuild of the chain never terminates).
+			name: "multi_pointer/named_pointer",
+			target: &struct {
+				P namedLeafPointer
+			}{},
+			exp: &struct {
+				P namedLeafPointer
+			}{
+				P: namedLeafPointer(ptrTo(pointerChainLeaf{Value: "v"})),
+			},
+			lookuper: MapLookuper(map[string]string{
+				"VALUE": "v",
+			}),
+		},
+		{
+			// A named pointer type in the middle of the chain.
+			name: "multi_pointer/pointer_to_named_pointer",
+			target: &struct {
+				P *namedLeafPointer
+			}{},
+			exp: &struct {
+				P *namedLeafPointer
+			}{
+				P: ptrTo(namedLeafPointer(ptrTo(pointerChainLeaf{Value: "v"}))),
+			},
+			lookuper: MapLookuper(map[string]string{
+				"VALUE": "v",
+			}),
+		},
+		{
+			// A named pointer-to-pointer type as the field type.
+			name: "multi_pointer/named_double_pointer",
+			target: &struct {
+				P namedDoubleLeafPointer
+			}{},
+			exp: &struct {
+				P namedDoubleLeafPointer
+			}{
+				P: namedDoubleLeafPointer(ptrTo(ptrTo(pointerChainLeaf{Value: "v"}))),
+			},
+			lookuper: MapLookuper(map[string]string{
+				"VALUE": "v",
 			}),
 		},
 

